Change app icon

malmstein opened the pull request:
Task/Issue URL: https://app.asana.com/0/72649045549333/1161651092205713

**Description**:
The iOS team saw a positive response from users after adding the feature that enabled users to change the app icon. We think that the same outcome can be achieved on the Android side.

**Disclaimer**:
It's known that this feature won't work on all devices. Depending on the launcher and OEM, it's possible that after selecting an icon it closes the app. If it doesn't, be aware that the icon change may take a few seconds before it's applied.

**Steps to test this PR**:
1. Open app 
2. Open Settings and Change App Icon
3. Select a new icon
4. Navigate to Apps Menu
5. Check new icon has been applied

**Screenshots**:
| Settings screen | Change Icon screen |
| ------ | ----- |
![device-2020-02-28-164625](https://user-images.githubusercontent.com/531613/75897758-5d51a500-5e39-11ea-8378-e93a200ee433.png)|![device-2020-03-04-103920 (1)](https://user-images.githubusercontent.com/531613/75897762-5e82d200-5e39-11ea-98d3-6d57b0d26ff1.png)

malmstein commented on app/src/main/AndroidManifest.xml:
```diff
@@ -11,7 +11,7 @@
         android:name="com.duckduckgo.app.global.DuckDuckGoApplication"
         android:allowBackup="false"
-        android:icon="@mipmap/ic_launcher"
+        android:icon="${appIcon}"
```
We can use manifest placeholders to change the app icon based on the build version. This is done to replace the need of maintaining a separate folder in the source.

malmstein commented on app/build.gradle:
```diff
@@ -45,11 +45,19 @@ android {
         debug {
             applicationIdSuffix ".debug"
             pseudoLocalesEnabled false
+            manifestPlaceholders = [
```
We can use manifest placeholders to change the app icon based on the build version. This is done to replace the need of maintaining a separate folder in the source.

CDRussell replied:
I'm no longer seeing the blue icon for DEBUG builds. Any thoughts why?

Pixel 2; Android R

malmstein replied:
This is now fixed, thanks for noticing!

CDRussell commented on app/src/androidTest/java/com/duckduckgo/app/icon/ui/ChangeIconViewModelTest.kt:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Test
+    fun storesIconWhenUserSelectsIt() {
```
Preferred test naming format is `whenXThenY()`

CDRussell commented on app/src/main/java/com/duckduckgo/app/di/VariantModule.kt:
```diff
@@ -35,4 +39,8 @@ class VariantModule {
     fun weightedRandomizer() = WeightedRandomizer()
+
+    @Provides
+    fun appIconModifier(context: Context, appShortcutCreator: AppShortcutCreator): IconModifier =
```
minor: this seems misplaced being in the `VariantModule`

malmstein replied:
I didn't want to create a Module just for this, probably `SystemComponentsModule` would be better

CDRussell commented on app/src/main/java/com/duckduckgo/app/icon/api/AppIconModifier.kt:
```diff
@@ -0,0 +1,108 @@
+        disable(context, previousIcon)
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
+            appShortcutCreator.configureAppShortcuts(context)
```
I'd been thinking it was good the app shortcuts changed icon for free, but see now you made it happen. 👍

CDRussell commented on app/src/main/java/com/duckduckgo/app/settings/db/SettingsDataStore.kt:
```diff
@@ -75,6 +77,13 @@ class SettingsSharedPreferences @Inject constructor(private val context: Context
+            val componentName = preferences.getString(KEY_APP_ICON, AppIcon.DEFAULT.componentName) ?: return AppIcon.DEFAULT
+            return AppIcon.from(componentName)
+        }
+        set(appIcon) = preferences.edit { putString(KEY_APP_ICON, appIcon.componentName) }
```
Worth trying `preferences.edit(commit = true)` here since we know the app might crash soon after? Maybe try ensure it has time to flush the changes to disk.

malmstein replied:
👍

CDRussell commented on app/src/main/java/com/duckduckgo/app/statistics/pixels/Pixel.kt:
```diff
@@ -155,7 +155,9 @@ interface Pixel {
-        AUTOCOMPLETE_SEARCH_SELECTION("m_aut_s_s")
+        AUTOCOMPLETE_SEARCH_SELECTION("m_aut_s_s"),
+
+        CHANGE_APP_ICON_OPENED("mic"),
```
We should be using `m_` prefix for all new pixels unless there's a good reason not to

malmstein replied:
Changed it to `m_ic` 👍

CDRussell commented:
<img width="989" alt="Screen Shot 2020-03-05 at 9 14 08 AM" src="https://user-images.githubusercontent.com/1336281/75966107-b9600c00-5ec1-11ea-8404-466b0f7761e6.png">

The amount of launcher icon assets we have now is huge. This might be the only way to do it, but thought I'd ask the question anyway to question assumptions: do we still need to provide all these `PNG` or could any of them be replaced with `XML` instead?

malmstein commented:
Unfortunately not @CDRussell. Adaptive icons are [only available API level 26 and above](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive). If and when we have minSDK 26 we can get rid of those pngs.

CDRussell commented on app/src/main/java/com/duckduckgo/app/fire/AutomaticDataClearer.kt:
```diff
@@ -171,13 +182,19 @@ class AutomaticDataClearer(
+    private fun shouldClearData(cleanWhenOption: ClearWhenOption, appUsedSinceLastClear: Boolean, appIconChanged: Boolean): Boolean {
         Timber.d("Determining if data should be cleared for option $cleanWhenOption")
 
+        if (appIconChanged) {
```
I like the way this is integrated into this class; feels like a natural place for it 👍

CDRussell commented on app/src/main/java/com/duckduckgo/app/launch/LaunchBridgeActivity.kt:
```diff
@@ -68,11 +68,11 @@ class LaunchBridgeActivity : DuckDuckGoActivity() {
             startActivity(OnboardingActivity.intent(this))
         }
-        finish()
+        // finish()
```
Can we kill these commented out lines? Also, if the intention is to delete it, why don't we call `finish()`?

malmstein replied:
This was just a test, good catch!

CDRussell commented on app/src/main/java/com/duckduckgo/app/settings/db/SettingsDataStore.kt:
```diff
@@ -85,6 +86,10 @@ class SettingsSharedPreferences @Inject constructor(private val context: Context
         }
         set(appIcon) = preferences.edit(commit = true) { putString(KEY_APP_ICON, appIcon.componentName) }
 
+    override var appIconChanged: Boolean
```
Could we add a comment here explaining this flag? i just know otherwise i'm going to forget it's a temporary flag just across the app restart

malmstein replied:
Yeah that make sense, I've added a comment explaining it

malmstein commented:
@CDRussell I've fixed the issue with the flag after app close. The tests were missing the flag `appUsedSinceLastClear` so they were green without having to be. I've added a few extra tests to make sure we are on the right side of things
